File: csharp/Platform.Protocols.Lino.Tests/MultilineStringParserTests.cs
using Xunit;

namespace Platform.Protocols.Lino.Tests
{
    public static class MultilineStringParserTests
    {
        [Fact]
        public static void MultilineDoubleQuotedStringTest()
        {
            var parser = new Parser();
            var source = @"(""long
string literal representing
the reference"")";
            var links = parser.Parse(source);
            Assert.Equal(1, links.Count);
            Assert.Equal(null, links[0].Id);
            Assert.Equal(1, links[0].Values.Count);
            Assert.Equal("long\nstring literal representing\nthe reference", links[0].Values[0].Id);
        }

        [Fact]
        public static void MultilineSingleQuotedStringTest()
        {
            var parser = new Parser();
            var source = @"('another
long string literal 
as another reference')";
            var links = parser.Parse(source);
            Assert.Equal(1, links.Count);
            Assert.Equal(null, links[0].Id);
            Assert.Equal(1, links[0].Values.Count);
            Assert.Equal("another\nlong string literal \nas another reference", links[0].Values[0].Id);
        }

        [Fact]
        public static void Issue53ExampleTest()
        {
            var parser = new Parser();
            // Test the exact example from issue #53
            var source = @"(
  ""long
string literal representing
the reference""
  
  'another
long string literal 
as another reference'
)";
            var links = parser.Parse(source);
            Assert.Equal(1, links.Count);
            Assert.Equal(null, links[0].Id);
            Assert.Equal(2, links[0].Values.Count);
            Assert.Equal("long\nstring literal representing\nthe reference", links[0].Values[0].Id);
            Assert.Equal("another\nlong string literal \nas another reference", links[0].Values[1].Id);
        }

        [Fact]
        public static void MultilineStringWithIdTest()
        {
            var parser = new Parser();
            var source = @"(myId: ""first
multiline
value"" 'second
multiline
value')";
            var links = parser.Parse(source);
            Assert.Equal(1, links.Count);
            Assert.Equal("myId", links[0].Id);
            Assert.Equal(2, links[0].Values.Count);
            Assert.Equal("first\nmultiline\nvalue", links[0].Values[0].Id);
            Assert.Equal("second\nmultiline\nvalue", links[0].Values[1].Id);
        }

        [Fact]
        public static void MixedSingleAndMultilineStringTest()
        {
            var parser = new Parser();
            var source = @"(normal ""multi
line"" single)";
            var links = parser.Parse(source);
            Assert.Equal(1, links.Count);
            Assert.Equal(null, links[0].Id);
            Assert.Equal(3, links[0].Values.Count);
            Assert.Equal("normal", links[0].Values[0].Id);
            Assert.Equal("multi\nline", links[0].Values[1].Id);
            Assert.Equal("single", links[0].Values[2].Id);
        }

        [Fact]
        public static void SingleCharacterMultilineStringTest()
        {
            var parser = new Parser();
            var source = @"(""a"")";
            var links = parser.Parse(source);
            Assert.Equal(1, links.Count);
            Assert.Equal(null, links[0].Id);
            Assert.Equal(1, links[0].Values.Count);
            Assert.Equal("a", links[0].Values[0].Id);
        }
    }
}

File: rust/tests/multiline_string_parser_tests.rs
use lino::{parse_lino_to_links, LiNo};

#[test]
fn multiline_double_quoted_string_test() {
    let source = r#"("long
string literal representing
the reference")"#;
    let links = parse_lino_to_links(source).unwrap();
    assert_eq!(links.len(), 1);
    
    match &links[0] {
        LiNo::Link { id, values } => {
            assert_eq!(id, &None);
            assert_eq!(values.len(), 1);
            if let LiNo::Ref(ref_id) = &values[0] {
                assert_eq!(ref_id, "long\nstring literal representing\nthe reference");
            } else {
                panic!("Expected Ref variant, got: {:?}", values[0]);
            }
        }
        LiNo::Ref(ref_id) => {
            // Single references can be optimized to direct Ref
            assert_eq!(ref_id, "long\nstring literal representing\nthe reference");
        }
    }
}

#[test]
fn multiline_single_quoted_string_test() {
    let source = r#"('another
long string literal 
as another reference')"#;
    let links = parse_lino_to_links(source).unwrap();
    assert_eq!(links.len(), 1);
    
    match &links[0] {
        LiNo::Link { id, values } => {
            assert_eq!(id, &None);
            assert_eq!(values.len(), 1);
            if let LiNo::Ref(ref_id) = &values[0] {
                assert_eq!(ref_id, "another\nlong string literal \nas another reference");
            } else {
                panic!("Expected Ref variant, got: {:?}", values[0]);
            }
        }
        LiNo::Ref(ref_id) => {
            // Single references can be optimized to direct Ref
            assert_eq!(ref_id, "another\nlong string literal \nas another reference");
        }
    }
}

#[test]
fn issue_53_example_test() {
    // Test the exact example from issue #53
    let source = r#"(
  "long
string literal representing
the reference"
  
  'another
long string literal 
as another reference'
)"#;
    let links = parse_lino_to_links(source).unwrap();
    assert_eq!(links.len(), 1);
    
    if let LiNo::Link { id, values } = &links[0] {
        assert_eq!(id, &None);
        assert_eq!(values.len(), 2);
        
        if let LiNo::Ref(ref_id1) = &values[0] {
            assert_eq!(ref_id1, "long\nstring literal representing\nthe reference");
        } else {
            panic!("Expected first value to be Ref variant");
        }
        
        if let LiNo::Ref(ref_id2) = &values[1] {
            assert_eq!(ref_id2, "another\nlong string literal \nas another reference");
        } else {
            panic!("Expected second value to be Ref variant");
        }
    } else {
        panic!("Expected Link variant");
    }
}

#[test]
fn multiline_string_with_id_test() {
    let source = r#"(myId: "first
multiline
value" 'second
multiline
value')"#;
    let links = parse_lino_to_links(source).unwrap();
    assert_eq!(links.len(), 1);
    
    if let LiNo::Link { id, values } = &links[0] {
        assert_eq!(id, &Some("myId".to_string()));
        assert_eq!(values.len(), 2);
        
        if let LiNo::Ref(ref_id1) = &values[0] {
            assert_eq!(ref_id1, "first\nmultiline\nvalue");
        } else {
            panic!("Expected first value to be Ref variant");
        }
        
        if let LiNo::Ref(ref_id2) = &values[1] {
            assert_eq!(ref_id2, "second\nmultiline\nvalue");
        } else {
            panic!("Expected second value to be Ref variant");
        }
    } else {
        panic!("Expected Link variant");
    }
}

#[test]
fn mixed_single_and_multiline_string_test() {
    let source = r#"(normal "multi
line" single)"#;
    let links = parse_lino_to_links(source).unwrap();
    assert_eq!(links.len(), 1);
    
    if let LiNo::Link { id, values } = &links[0] {
        assert_eq!(id, &None);
        assert_eq!(values.len(), 3);
        
        if let LiNo::Ref(ref_id1) = &values[0] {
            assert_eq!(ref_id1, "normal");
        } else {
            panic!("Expected first value to be Ref variant");
        }
        
        if let LiNo::Ref(ref_id2) = &values[1] {
            assert_eq!(ref_id2, "multi\nline");
        } else {
            panic!("Expected second value to be Ref variant");
        }
        
        if let LiNo::Ref(ref_id3) = &values[2] {
            assert_eq!(ref_id3, "single");
        } else {
            panic!("Expected third value to be Ref variant");
        }
    } else {
        panic!("Expected Link variant");
    }
}

#[test]
fn single_character_multiline_string_test() {
    let source = r#"("a")"#;
    let links = parse_lino_to_links(source).unwrap();
    assert_eq!(links.len(), 1);
    
    
    match &links[0] {
        LiNo::Link { id, values } => {
            assert_eq!(id, &None);
            assert_eq!(values.len(), 1);
            if let LiNo::Ref(ref_id) = &values[0] {
                assert_eq!(ref_id, "a");
            } else {
                panic!("Expected Ref variant, got: {:?}", values[0]);
            }
        }
        LiNo::Ref(ref_id) => {
            // It might be parsed as a simple reference instead of a link
            assert_eq!(ref_id, "a");
        }
    }
}